discovery-provider/src/challenges/challenge.py

```diff
@@ -1,41 +1,75 @@
 import logging
+from typing import Counter, Dict, Tuple, TypedDict, List, Optional, cast
+from sqlalchemy.orm.session import Session
+from sqlalchemy import func
+from src.models.models import ChallengeType
 from abc import ABC, abstractmethod
 from src.models import Challenge, UserChallenge
 
 logger = logging.getLogger(__name__)
 
 # DB Accessors
-def fetch_user_challenges(session, challenge_id, user_ids):
+def fetch_user_challenges(
+    session: Session, challenge_id: str, specifiers: List[str]
+) -> List[UserChallenge]:
     return (
         session.query(UserChallenge).filter(
             UserChallenge.challenge_id == challenge_id,
-            UserChallenge.user_id.in_(user_ids),
+            UserChallenge.specifier.in_(specifiers),
         )
     ).all()
 
 
+class EventMetadata(TypedDict):
+    block_number: int
+    user_id: int
+    extra: Dict
+
+
+class FullEventMetadata(TypedDict):
+    block_number: int
+    user_id: int
+    specifier: str
+    extra: Dict
+
+
 class ChallengeUpdater(ABC):
     """`ChallengeUpdater` is an abstract class which provides challenge specific logic
     to an instance of a `ChallengeManager`. The only required override is update_user_challenges
     """
 
-    @abstractmethod
     def update_user_challenges(
-        self, session, event, user_challenges_metadata, step_count
+        self,
+        session: Session,
+        event: str,
+        user_challenges: List[UserChallenge],
```

Comment:
Just want to make sure that the `update_user_challenges` method will not need the `extra` field from the event metadata.

Reply:
I don't think it will, since stuff in extra should be baked into the UserChallenge or in the DB state this point. Valid callout tho, possible we'll need it later. Should be easy to add if needed